### PR TITLE
EG-2754, EG-2755 Modify Additional Identity flow implementation

### PR DESCRIPTION
**Pull request by tubibuto**
## Description

This is the second PR in the second phase of Business Networks membership management implementation which takes care of implementation of additional identities support.

First we create new `MembershipIdentity` data class which consists of required `cordaIdentity` and optional `additionalIdentity` field. Additional identity field is represented by `BNIdentity` interface which every custom defined additional identity class should implement. We also add another `AdminPermission` named `CAN_MODIFY_ADDITIONAL_IDENTITY`.

After that we implement associated contract logic inside `MembershipContract` by adding `verifyModifyAdditionalIdentity()` open method specialised for verification of `ModifyAdditionalIdentity` command. All the supporting states and contract tests can be found in `MembershipStateTest` and `MembershipContractTest`.

Finally we get to jump into flow writing. We modify `CreateBusinessNetworkFlow` and `RequestMembershipFlow` so that they take optional `additionalIdentity` argument. This way it is possible for requesting members to provide additional BN specific identity by which authorised members can decide whether they want to activate given pending membership. 

Now we introduce `ModifyAdditionalIdentityFlow` which takes `membershipId` of membership to modify and `additionalIdentity` to assign to that membership. This is the workflow:
- Check initiator is authorised to modify additional identity
- Initiator builds additional identity modification transaction and sends it for signing to all active member authorised to modify membership
- Transaction is finalised and sent to all observers who are initiator, modified member and all members authorised to modify membership

In the end we add test coverage inside `ModifyAdditionalIdentityFlowTest` and amend `MembershipManagementFlowTest` to support more test helper methods. We also add additional identity checks in existing membership management tests.

## Test Plan

Ensure all existing and new Business Network related tests pass.

Deploy nodes, create Business Network with 2 active members and try to modify additional identity on them.

BNO modifying Member1's additional identity:
Screenshot 2020-06-29 at 19 03 49
Screenshot 2020-06-29 at 19 04 06

### CODE REVIEW

File: business-networks/workflows/src/main/kotlin/net/corda/bn/flows/ModifyRolesFlow.kt [hunk 1]
```diff
@@ -44,8 +44,8 @@ class ModifyRolesFlow(private val membershipId: UniqueIdentifier, private val ro
 
         // fetch observers and signers
         val authorisedMemberships = databaseService.getMembersAuthorisedToModifyMembership(networkId).toSet()
-        val observers = authorisedMemberships.map { it.state.data.identity }.toSet() + membership.state.data.identity - ourIdentity
-        val signers = authorisedMemberships.filter { it.state.data.isActive() }.map { it.state.data.identity } - membership.state.data.identity
+        val observers = authorisedMemberships.map { it.state.data.identity.cordaIdentity }.toSet() + membership.state.data.identity.cordaIdentity - ourIdentity
+        val signers = authorisedMemberships.filter { it.state.data.isActive() }.map { it.state.data.identity.cordaIdentity } - membership.state.data.identity.cordaIdentity
```
Comment (johnbuckle) [inline business-networks/workflows/src/main/kotlin/net/corda/bn/flows/ModifyRolesFlow.kt:48]: Curious, is it ever possible for all the states to be inactive, hence there are no signers?

**Review by johnbuckle — APPROVED** [review]

File: business-networks/workflows/src/main/kotlin/net/corda/bn/flows/ModifyRolesFlow.kt (hunk 1 above)
Comment (tubibuto) [reply business-networks/workflows/src/main/kotlin/net/corda/bn/flows/ModifyRolesFlow.kt:48]: Right now it's impossible to suspend/revoke yourself or strip yourself of permissions so this means there will always be an active member inside business network

File: business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt [hunk 2]
```diff
@@ -25,6 +26,7 @@ open class MembershipContract : Contract {
         class Suspend(requiredSigners: List<PublicKey>) : Commands(requiredSigners)
         class Revoke(requiredSigners: List<PublicKey>) : Commands(requiredSigners)
         class ModifyRoles(requiredSigners: List<PublicKey>, val bnCreation: Boolean = false) : Commands(requiredSigners)
+        class ModifyAdditionalIdentity(requiredSigners: List<PublicKey>, val initiator: Party) : Commands(requiredSigners)
```
Comment (bpaunescu) [inline business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt]: Why is it called "additional identity" ? How about something like identity or business identity?

File: business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt [hunk 3]
```diff
@@ -87,8 +90,9 @@ open class MembershipContract : Contract {
         "Input state of membership activation transaction shouldn't be already active" using (!inputMembership.isActive())
         "Output state of membership activation transaction should be active" using (outputMembership.isActive())
         "Input and output state of membership activation transaction should have same roles set" using (inputMembership.roles == outputMembership.roles)
+        "Input and output state of membership activation transaction should have same additional identity" using (inputMembership.identity.additionalIdentity == outputMembership.identity.additionalIdentity)
```
Comment (bpaunescu) [inline business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt]: Sorry for grammar nitpicking: input and output state**s**.  Needs to be changed in several places now that I read carefully. Happy to make a note of it and change later.

**Review by bpaunescu — COMMENTED** [review]
Looks ok. Some comments in-line. I'm not fond of name "additional identity". Made some suggestions in the comments.

File: business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt (hunk 2 above)
Comment (tubibuto) [reply business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt]: I like business identity, so we'd have identity = corda identity + business identity

File: business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt (hunk 3 above)
Comment (tubibuto) [reply business-networks/contracts/src/main/kotlin/net/corda/bn/contracts/MembershipContract.kt]: What do you say we do a quick grammar scan and fix in the end of phase 2 - when relationships are sorted ?

**Pull request merged**